extensions/amp-sidebar/0.1/swipe-to-dismiss.js

```diff
@@ -0,0 +1,355 @@
+
+    return this.mutateElement_(() => {
+      setStyles(dev().assertElement(this.swipeElement_), {
+        transform: this.translateBy_(0),
+        transition: `${duration}ms transform ease-in`,
+      });
+      setStyles(dev().assertElement(this.mask_), {
+        opacity: '',
+        transition: `${duration}ms opacity ease-in`,
+      });
+    }).then(() => {
+      return delayAfterDeferringToEventLoop(this.win_, duration);
+    });
+  }
+
+  /**
+   * Adjusts the UI elements for the current swipe position in a swipe to
+   * dismiss gesture. This should be called in a mutate context.
+   * @param {string} swipeElementTransform How to transform the swiping
+   *    element.
+   * @param {number|string} maskOpacity The opacity for the mask element.
+   *    container.
+   * @private
+   */
+  adjustForSwipePosition_(swipeElementTransform = '', maskOpacity = '') {
+    setStyles(dev().assertElement(this.swipeElement_), {
+      transform: swipeElementTransform,
+      transition: '',
+    });
+    setStyles(dev().assertElement(this.mask_), {
+      opacity: maskOpacity,
+      transition: '',
+    });
+  }
+
+  /**
+   * Releases the user's swipe to dismiss gesture. This carries the momentum
+   * forwards and either closes the lightbox or snaps back based on the speed
+   * and distance. This should be called in a mutate context.
+   * @param {number} velocityX The X velocity when the swipe was released.
+   * @param {number} velocityY The Y velocity when the swipe was released.
+   * @param {number} deltaX The x distance when the swipe was released.
+   * @param {number} deltaY The y distance when the swipe was released.
+   * @return {!Promise} A Promise that resolves once the release is completed,
+   *    either snapping back to the start or closing the carousel.
+   * @private
+   */
+  releaseSwipe_(velocityX, velocityY, deltaX, deltaY) {
+    const distanceX = velocityX * SWIPE_TO_CLOSE_VELOCITY_TO_DISTANCE_FACTOR;
+    const distanceY = velocityY * SWIPE_TO_CLOSE_VELOCITY_TO_DISTANCE_FACTOR;
+    const finalDeltaX = distanceX + deltaX;
+    const finalDeltaY = distanceY + deltaY;
+    // We want to figure out the final distance we will rest at if the user
+    // flicked the element and use that to determine we should animate to. We
+    // will then use that resting position to determine if we should snap back
+    // or close.
+    const finalDistance = this.capDistance_(finalDeltaX, finalDeltaY);
+    const finalVelocity = this.capDistance_(velocityX, velocityY);
+
+    const swipeToCloseDistance =
+      this.getSwipeElementLength_() * SWIPE_TO_CLOSE_DISTANCE_THRESHOLD;
+    if (
+      finalDistance < swipeToCloseDistance &&
+      finalVelocity < SWIPE_TO_CLOSE_VELOCITY_THRESHOLD
+    ) {
+      return this.carrySwipeMomentum_(finalDistance, finalVelocity).then(() =>
+        this.snapBackFromSwipe_(finalDistance)
+      );
+    }
+
+    const remainingDistance = this.getSwipeElementLength_() - finalDistance;
```

Comment:
I think this would be a bit cleaner if we pull this logic out into a separate method like `dismissFromSwipe_` (or something like that).